File: Knapsack with Duplicate Items - GFG/knapsack-with-duplicate-items.js
//{ Driver Code Starts
//Initial Template for javascript


'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let [N,W] =  readLine().trim().split(" ").map((x) => parseInt(x));
        let val =  readLine().trim().split(" ").map((x) => parseInt(x));
        let wt =  readLine().trim().split(" ").map((x) => parseInt(x));
        let obj = new Solution();
        let res = obj.knapSack(N,W,val,wt);
        if(res === -0)
            res = 0;
        console.log(res);
        
    }
}
// } Driver Code Ends


//User function Template for javascript

/**
 * @param {number} N
 * @param {number} W
 * @param {number[]} val
 * @param {number[]} wt
 * @return {number}
*/

class Solution {

    knapSack(N,W,val,wt){
       // initialize dp array
        let dp = new Array(W+1).fill(0);
        let ans = 0;

        for(let i=0; i<W+1; i++) {
            for (let j=0; j<N; j++) {
                if(wt[j] <= i){
                    dp[i] = Math.max(dp[i], dp[i - wt[j]] + val[j]);
                }
            }
        }
        
        return dp[W];
    }
}

File: Knapsack with Duplicate Items - GFG/knapsack-with-duplicate-items.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG{
    public static void main(String args[])throws IOException
    {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(in.readLine());
        while(t-- > 0){
            String line1[] = in.readLine().trim().split("\\s+");
            int N = Integer.parseInt(line1[0]);
            int W = Integer.parseInt(line1[1]);
            String line2[] = in.readLine().trim().split("\\s+");
            int val[] = new int[N];
            for(int i = 0;i < N;i++)
                val[i] = Integer.parseInt(line2[i]);
            String line3[] = in.readLine().trim().split("\\s+");
            int wt[] = new int[N];
            for(int i = 0;i < N;i++)
                wt[i] = Integer.parseInt(line3[i]);
                
            Solution ob = new Solution();
            System.out.println(ob.knapSack(N, W, val, wt));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution{
    static int knapSack(int N, int W, int val[], int wt[])
    {
        // Creating an array to store the maximum value for each weight
        int dp[] = new int[W+1];
		int ans =0;

		for(int i=0; i<W+1; i++) // Iterating through all possible weights
			for (int j=0; j<N; j++) // Iterating through all items
				if(wt[j]<= i) // If the weight of the item is less than or equal to the current weight
					dp[i] = Math.max(dp[i], dp[i - wt[j]] + val[j]); // Updating the maximum value for the current weight
		return dp[W]; // Returning the maximum value for the total weight
    }
}